### ISSUE TITLE
Support afterRender hook for custom elements/attributes

### ISSUE THREAD

**Issue by shanonvl** [aurelia/templating#79]
I have a custom element 
```
<x-foo bar.bind="[1,2]"></x-foo>
```
which has a template:
```
<template>
   <ul>
      <li repeat.for="f of foo">${f}</li>
   </ul>
</template>
```
and resulting HTML:
```
<ul>
<li>1</li><li>2</li>
</ul>
```

I have a `fooChanged` event which allows me to hook in *before* the DOM has been updated with the new `<li>`'s for the updated `foo`:
```
### something sets foo = [1,2,3] ###
---> fooChanged(newValue,oldValue)
---> HTML updated:
<ul>
<li>1</li><li>2</li><li>3</li>
</ul>
```

Is there a hook for *after* the DOM has been updated instead of before? 

e.g. 
```
### something sets foo = [1,2,3] ###
---> fooChanged(newValue,oldValue)
---> HTML updated:
<ul>
<li>1</li><li>2</li><li>3</li>
</ul>
--> afterRender(...)
```
Please let me know if I can provide any other info or if there's already a way to do this in Aurelia.  Thanks!

**Comment by shanonvl** [comment]
Note that I use the `attached` handler to do the for the *first* render...

**Comment by EisenbergEffect** [comment]
Can you explain what it is you want to accomplish with this? We might have a way to do it already, or if not, your use case can help use determine specific details for our implementation.

**Comment by shanonvl** [comment]
Absolutely.   I'm building a data table using the jQuery Data Tables plugin.   I'm currently trying to generate much of the markup using templating ( configs via `data-` and columns via `thead.tr.th`'s and  and then, once the dom is rendered, instantiate `$('table').DataTable(...)`  When the model changes, the `changed` handlers are executing before the underlying HTML is updated so if I'm in the `columnsChanged(newValue,oldValue)` handler, the `thead.tr.th`'s represent `oldValue` vs. `newValue`.

I've come up with a few workarounds for this, so this isn't a big issue for me but I would use the functionality if present.   Thanks!

**Comment by EisenbergEffect** [comment]
Can you share a bit about what workarounds you are using currently?

**Comment by shanonvl** [comment]
Well, workarounds are really just a single workaround:  move to fully JS-based rendering.  Use a single config property from the parent view model which is bound to the custom element e.g. `<x-data-table value.bind="tableConfig"`></x-data-table>` and use the `valueChanged` to re-render the content using JS entirely.

I'd prefer to do it with discretely bound objects so that I can more clearly factor my custom element class, however this method removes the hard requirement for an `afterRender`-like hook.

**Comment by EisenbergEffect** [comment]
You might want to look at using the micro task queue. Perhaps this will be
of use?
https://github.com/aurelia/templating-resources/blob/master/src/compose.js

On Sun, May 24, 2015 at 2:10 PM, Shanon Levenherz <notifications@github.com>
wrote:

> Well, workarounds are really just a single workaround: move to fully
> JS-based rendering. Use a single config property from the parent view model
> which is bound to the custom element e.g. <x-data-table
> value.bind="tableConfig">and use thevalueChanged` to re-render the
> content using JS entirely.
>
> I'd prefer to do it with discretely bound objects so that I can more
> clearly factor my custom element class, however this method removes the
> hard requirement for an afterRender-like hook.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/aurelia/templating/issues/79#issuecomment-105007460>.
>



-- 
Rob Eisenberg,
President - Blue Spire
www.durandaljs.com

**Comment by shanonvl** [comment]
Thanks @EisenbergEffect - `queueMicroTask` is working well for me.   Really appreciate the assistance here.

**Comment by EisenbergEffect** [comment]
You're welcome. I'll keep this open so we can evaluate if we need to add something else in the future or at least do some education around this scenario.

**Comment by shanonvl** [comment]
Great -- please let me know if I can provide any other info/examples; I'd be glad to help however possible.

**Comment by buzinas** [comment]
I had the same needs for building Google MDL components, and I guess everyone who needs to call some wrapper after the DOM re-rendering will come through this

When I had this problems, people suggested me some workarounds ultil Rob went to the Gitter channel and also told me to use the queueMicroTask, and it worked smoothly.

But since this is something that's highly needed to integrate with most of the 3rd-party UIs, I think there should be an global event for that, like there is "bind", "activate", "attached". The OP suggested "afterRender", and on Gitter I have suggested "domUpdated". Not sure about what name would be the best, but for sure it would be nice to have such a feature.

**Comment by stoffeastrom** [comment]
As I wrote on gitter https://gitter.im/Aurelia/Discuss?at=55ba9cbb7978296537a2bd8f. You actually need to queue it on the macro task queue since the UI rendering is happening there. That's basically doing a `setTimeout`. That said I would be very careful adding that kind of event for every view since eventually it could be a real performance killer. So if you don't want that event you don't want that penalty. The framework could perhaps detect if that event is defined and queue it on the macro task queue?

**Comment by EisenbergEffect** [comment]
Yep, you can use the micro task queue for this :)

**Comment by nexbit** [comment]
Hi all
sorry for bothering you, but can you point out a comprehensive article/example on how to use this so much talked about micro task queue in Aurelia? I can't find anything complete on the web.
I'm trying to solve some problems with third-party jquery components (please please Aurelia Interface soooon ;)), and an "after render" hook would come in handy.
Thanks!

**Comment by EisenbergEffect** [comment]
```
import {inject, TaskQueue} from 'aurelia-framework';

@inject(TaskQueue)
export class MyElement {
  constructor(taskQueue) {
    this.taskQueue = taskQueue;
  }

  scheduleSomething() {
    this.taskQueue.queueMicroTask(() => doSomething());
  }
}
```

Without more details, I don't know if this will solve your specific problem, but that is how you use the Micro Task Queue. You can see the api here: http://aurelia.io/docs.html#/aurelia/task-queue/1.0.0-beta.1.0.1/doc/api/class/TaskQueue

**Comment by nexbit** [comment]
@EisenbergEffect , that was sufficient..
I supposed there were much more code and trickery involved, but now I've just read the task-queue code and I know exactly when/where using it. ;) Indeed, I was already using mutation observers in other modules, but didn't realize they could be used to schedule tasks after DOM updates (but it makes a lot of sense now!).

A side question: is it possible/advisable to extend the default view lifecycle to provide custom extension points, like the aforementioned afterRender hook?

Thank you very much for pointing me in the right direction!
P.

**Comment by EisenbergEffect** [comment]
I don't see that there's an issue for you to standardize on that sort of hook in your app, if you need it. However, I'm reluctant to extend Aurelia's lifecycle beyond what it is for now.

**Comment by pasky** [comment]
My story: Generating <img>s from an array and wanting to use imagesLoaded+masonry - so I need to invoke a javascript snippet only after I'm sure the <img>s are in the DOM. queueMicroTask did *not* work for me, but queueTask seems to.
